utils/zerovec/src/varzerovec/components.rs

```diff
@@ -0,0 +1,246 @@
+                let start = usizeify(win[0]);
+                let end = usizeify(win[1]);
+                unsafe { self.things.get_unchecked(start..end) }
+            })
+            .chain(last)
+            .map(|s| unsafe { T::VarULE::from_byte_slice_unchecked(s) })
+    }
+}
+
+impl<'a, T> SliceComponents<'a, T>
+where
+    T: AsVarULE,
+    T::VarULE: Ord,
+{
+    pub fn binary_search(&self, x: &T::VarULE) -> Result<usize, usize> {
+        // This code is an absolute atrocity. This code is not a place of honor. This
+        // code is known to the State of California to cause cancer.
+        //
+        // Unfortunately, the stdlib's `binary_search*` functions can only operate on slices.
+        // We do not have a slice. We have something we can .get() and index on, but that is not
+        // a slice.
+        //
+        // The `binary_search*` functions also do not have a variant where they give you the element's
+        // index, which we could otherwise use to directly index `self`.
+        // We do have `self.indices`, but these are indices into a byte buffer, which cannot in
+        // isolation be used to recoup the logical index of the element they refer to.
+        //
+        // However, `binary_search_by()` provides references to the elements of the slice being iterated.
+        // Since the layout of Rust slices is well-defined, we can do pointer arithmetic on these references
+        // to obtain the index being used by the search.
+        //
+        // It's worth noting that the slice we choose to search is irrelevant, as long as it has the appropriate
+        // length. `self.indices` is defined to have length `self.len()`, so it is convenient to use
+        // here and does not require additional allocations.
+        //
+        // The alternative to doing this is to implement our own binary search. This is significantly less fun.
+
+        let zero_index = self.indices.as_ptr() as *const _ as usize;
+        self.indices.binary_search_by(|probe: &_| {
+            // `self.indices` is a vec of unaligned u32s, so we divide by sizeof(u32)
+            // to get the actual index
+            let index = (probe as *const _ as usize - zero_index) / mem::size_of::<u32>();
+            // safety: we know this is in bounds
+            let actual_probe = unsafe { self.get_unchecked(index) };
+            actual_probe.cmp(x)
+        })
+    }
+}
+
+pub fn get_serializable_bytes<T: AsVarULE>(elements: &[T]) -> Option<Vec<u8>> {
+    let mut vec = Vec::new();
+    let len_u32: u32 = elements.len().try_into().ok()?;
+    vec.extend(&len_u32.as_unaligned().0);
+    let mut offset: usize = 0;
+    for element in elements {
+        let offset_u32: u32 = offset.try_into().ok()?;
+        vec.extend(&offset_u32.as_unaligned().0);
+        offset += element.as_unaligned().as_byte_slice().len();
+    }
+    for element in elements {
+        vec.extend(element.as_unaligned().as_byte_slice())
+    }
```

Comment:
`suggestion(optional)`: could you use `flatten` here on iter/collect?

Reply:
I instead just reserved the right amount of space, I don't think `flatten` will help that much here and it can get a little bit confusing as to what is being appended (even though the types _should_ help)